vaadin-upload-flow-parent/vaadin-upload-flow/src/main/java/com/vaadin/flow/component/upload/Upload.java

```diff
@@ -653,6 +675,37 @@ public void setReceiver(Receiver receiver) {
         } else {
             setMaxFiles(1);
         }
+
+        runBeforeClientResponse(ui -> getElement().setAttribute("target",
+                new StreamReceiver(getElement().getNode(), "upload",
+                        getStreamVariable())));
+    }
+
+    /**
+     * Set the upload handler to be used for this upload component.
+     * <p>
+     * The given handler defines how uploaded file content is handled on the
+     * server and invoked per each single file to be uploaded. Note! This method
+     * overrides the receiver set by {@link #setReceiver(Receiver)} and removes
+     * the maximum files limit.
+     *
+     * @param handler
+     *            upload handler to use for file receptions
+     */
+    public void setUploadHandler(UploadHandler handler) {
+        AbstractStreamResource elementStreamResource = new StreamResourceRegistry.ElementStreamResource(
+                handler, this.getElement());
+
+        runBeforeClientResponse(ui -> getElement().setAttribute("target",
+                elementStreamResource));
+
+        if (handler instanceof TransferProgressAwareHandler<?, ?> transferProgressAwareHandler) {
+            transferProgressAwareHandler.addTransferProgressListener(
+                    new UploadTransferProgressListener(this));
+        }
```

Comment:
How is this handled for user generated UploadHandlers?
The events seem to be sort of crucial to the component state management

Reply:
Excellent catch! 
I wonder if I can replace the TPL with a simple wrapping for all cases:
```
UploadHadnler newUploadHandler = event -> {
                try {
                    startUpload();
                    handler.handleUploadRequest(event);
                } finally {
                    endUpload();
                }
            };
```